lib/temporal/worker.rb

```diff
@@ -0,0 +1,72 @@
+require 'async'
+require 'temporal/bridge'
+require 'temporal/data_converter'
+require 'temporal/failure_converter'
+require 'temporal/payload_converter'
+require 'temporal/runtime'
+require 'temporal/worker/activity'
+require 'temporal/worker/thread_pool_executor'
+
+module Temporal
+  class Worker
+    # TODO: Add worker interceptors
+    def initialize(
+      connection,
+      namespace,
+      task_queue,
+      payload_converter: Temporal::PayloadConverter::DEFAULT,
+      payload_codecs: [],
+      failure_converter: Temporal::FailureConverter::DEFAULT,
+      activity_executor: nil,
+      max_concurrent_activities: 100
+    )
+      @running = false
+      @mutex = Mutex.new
+      @runtime = Temporal::Runtime.instance
+      activity_executor ||= ThreadPoolExecutor.new(max_concurrent_activities)
+      data_converter = DataConverter.new(
+        payload_converter: payload_converter,
+        payload_codecs: payload_codecs,
+        failure_converter: failure_converter,
+      )
+      core_worker = Temporal::Bridge::Worker.create(
+        @runtime.core_runtime,
+        connection.core_connection,
+        namespace,
+        task_queue,
+      )
+      @activity_worker = Worker::Activity.new(core_worker, data_converter, activity_executor)
+      @workflow_worker = nil
+    end
+
+    def run
+      Async { |task| start(task) }
```

Comment:
This will block a thread whereas start will not correct?

I think that's probably ok, but we need some kind of "non-thread-blocking run that still raises exceptions". Fatal errors can happen in workers and we need to raise them to users. Ideally they don't have to register some kind error handler or just count on logging.

Can we do some kind of "run" that accepts a block of code and raise inside of that code?

This has bit us in a couple of SDKs now where we did the "start-and-return" paradigm only to not have anywhere to throw fatal errors to the user. We'd rather _they_ made a non-blocking start than us.

Reply:
> I'm not sure what might be a good solution

Can we just offer the blocking call and let the user choose how to make it async? Then it's up to them how to handle errors thrown from the blocking call.

> Alternatively we can call their block (if provided), but that would only happen on the Reactor thread, which might not be the main one

Yeah, I no longer thing that is the best way. Ignore I suggested accepting a block.

Reply:
Looks like we decided against this in our slack thread.